### PR TITLE
rework way how we configure service providers

### PR DESCRIPTION
**Pull request by sparkoo**
### What does this PR do?
Initialize one controller per service provider type. The controller holds configuration for individual service providers of the type based on baseurl.

This enable us to have multiple service provider configurations of the same type and to use user oauth configuration even in case url of serviceprovider url is not known.

Changes spi oauth url from `/<service_provider>/authenticate` to `/oauth/authenticate`

TODO: tests

### What issues does this PR fix or reference?

https://issues.redhat.com/browse/SVPI-249

### How to test this PR?

images: 
`quay.io/mvala/service-provider-integration-operator:svpi249-oauthconfigs`
`quay.io/mvala/service-provider-integration-oauth:svpi249-oauthconfigs`

 - current global and user configuration must work
 - try multiple same SP in global configuration like:
```
serviceProviders:
- type: GitLab
  clientId: "aaa"
  clientSecret: "111"
  baseUrl: "https://gitlab.cz"
- type: GitLab
  clientId: "123"
  clientSecret: "234"
  baseUrl: "https://blabol.com"
```
 - user config secret can now have `spi.appstudio.redhat.com/service-provider-host` label, defining host of service provider:
```
apiVersion: v1
kind: Secret
type: Opaque
metadata:
  labels:
    spi.appstudio.redhat.com/service-provider-type: GitLab
    spi.appstudio.redhat.com/service-provider-host: gitlaaab.com
  name: sc1
  namespace: default
data:
  clientId: c2MxYw==
  clientSecret: c2Mxcw==
```

### CODE REVIEW

**Review by mshaposhnik — APPROVED** [review]
I have checked this in pair with file retriever and no problems were met.
The only proposal is to indidate more clearly that we rework specifically the OAuth part, because reading description a first time gave me impression that we're reworking the CRD controllers :-), i was scared.

**Review by xbaran4 — APPROVED** [review]
Works for me.

image
image

File: oauth/oauth_config.go [hunk 1]
```diff
@@ -144,13 +159,20 @@ func initializeConfigFromSecret(secret *corev1.Secret, oauthCfg *oauth2.Config)
...
 	}
 
-	if tokenUrl, has := secret.Data[oauthCfgSecretFieldTokenUrl]; has {
+	if tokenUrl, has := secret.Data[oauthCfgSecretFieldTokenUrl]; has && len(tokenUrl) > 0 {
 		oauthCfg.Endpoint.TokenURL = string(tokenUrl)
 	}
 
 	return nil
 }
+
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/oauth_config.go:172]: this is not necessarily for saas. it's used for any url in the config.

File: oauth/github.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oauth
+
+var GithubUrlBaseHost = "github.com"
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/github.go]: fixed

File: oauth/gitlab.go [hunk 3]
```diff
@@ -17,8 +17,10 @@ import (
...
 
-// gitlabEndpoint is the OAuth endpoints specification of SAAS GitLab instance.
-var gitlabEndpoint = oauth2.Endpoint{
+// GitlabEndpoint is the OAuth endpoints specification of SAAS GitLab instance.
+var GitlabEndpoint = oauth2.Endpoint{
 	AuthURL:  "https://gitlab.com/oauth/authorize",
 	TokenURL: "https://gitlab.com/oauth/token",
 }
+
+var GitlabUrlBaseHost = "gitlab.com"
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/gitlab.go]: fixed

File: oauth/oauth_config.go [hunk 4]
```diff
@@ -40,8 +40,9 @@ const (
 )
 
 var (
-	errMissingField            = errors.New("missing mandatory field in oauth configuration")
-	errMultipleMatchingSecrets = errors.New("found multiple matching oauth config secrets")
+	errMissingField                     = errors.New("missing mandatory field in oauth configuration")
+	errMultipleMatchingSecrets          = errors.New("found multiple matching oauth config secrets")
+	errConfigNotFoundForServiceProvider = errors.New("haven't found oauth configuration for service provider")
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/oauth_config.go]: fixed

File: oauth/oauth_config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/oauth_config.go:172]: eventually we can, but definitely not as a part of this PR. I think well-known url is not guaranteed at all and we would have to do it per service provider.

File: oauth/oauth_config.go [hunk 5]
```diff
@@ -67,20 +84,22 @@ func (c *commonController) obtainOauthConfig(ctx context.Context, info *oauthsta
...
-	} else {
+	} else if foundDefaultOauthConfig {
 		lg.V(logs.DebugLevel).Info("using default oauth config")
-		oauthCfg.ClientID = c.Config.ClientId
-		oauthCfg.ClientSecret = c.Config.ClientSecret
+		oauthCfg.ClientID = defaultOauthConfig.Config.ClientId
+		oauthCfg.ClientSecret = defaultOauthConfig.Config.ClientSecret
 		return oauthCfg, nil
+	} else {
+		return nil, fmt.Errorf("%w '%s' url: '%s'", errConfigNotFoundForServiceProvider, info.ServiceProviderType, info.ServiceProviderUrl)
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/oauth_config.go]: fixed

File: oauth/controller.go [hunk 6]
```diff
@@ -50,43 +48,59 @@ const (
...
-				AuthURL:  spConfig.ServiceProviderBaseUrl + "/oauth/authorize",
-				TokenURL: spConfig.ServiceProviderBaseUrl + "/oauth/token",
+	for _, sp := range cfg.ServiceProviders {
+		if sp.ServiceProviderType != spType {
+			continue
+		}
+
+		baseUrl := defaultBaseUrlHost
+		if sp.ServiceProviderBaseUrl != "" {
+			baseUrlParsed, parseUrlErr := url.Parse(sp.ServiceProviderBaseUrl)
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/controller.go:82]: fixed

File: cmd/oauth/oauth.go [hunk 7]
```diff
@@ -132,36 +134,61 @@ func main() {
...
+	routerCfg := oauth.RouterConfiguration{
+		OAuthServiceConfiguration: cfg,
+		Authenticator:             authenticator,
+		StateStorage:              stateStorage,
+		K8sClient:                 cl,
+		TokenStorage:              strg,
+		RedirectTemplate:          redirectTpl,
+	}
+	// all servise provider types we support, including default values
+	serviceProviderDefaults := []oauth.ServiceProviderDefaults{
```
[the remark this answers is not in the document]
Comment (sparkoo) [reply cmd/oauth/oauth.go:163]: I've commented the https://issues.redhat.com/browse/SVPI-271 issue

File: oauth/oauth_config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sparkoo) [reply oauth/oauth_config.go:169]: issue created https://issues.redhat.com/browse/SVPI-274

**Pull request merged**